server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -0,0 +1,182 @@
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.union;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.function.Supplier;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobStore;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.blob.api.ObjectNotFoundException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+
+public class HybridBlobStore implements BlobStore {
+    @FunctionalInterface
+    public interface RequireLowCost {
+        RequirePerforming lowCost(BlobStore blobStore);
+    }
+
+    @FunctionalInterface
+    public interface RequirePerforming {
+        Builder performing(BlobStore blobStore);
+    }
+
+    public static class Builder {
+        private final BlobStore lowCostBlobStore;
+        private final BlobStore legacyBlobStore;
+
+        Builder(BlobStore lowCostBlobStore, BlobStore legacyBlobStore) {
+            this.lowCostBlobStore = lowCostBlobStore;
+            this.legacyBlobStore = legacyBlobStore;
+        }
+
+        public HybridBlobStore build() {
+            return new HybridBlobStore(
+                lowCostBlobStore,
+                legacyBlobStore);
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HybridBlobStore.class);
+    private static final int SIZE_THRESHOLD = 32 * 1024;
+
+    public static RequireLowCost builder() {
+        return lowCost -> performing -> new Builder(lowCost, performing);
+    }
+
+    private final BlobStore lowCostBlobStore;
+    private final BlobStore performingBlobStore;
+
+    private HybridBlobStore(BlobStore lowCostBlobStore, BlobStore performingBlobStore) {
+        this.lowCostBlobStore = lowCostBlobStore;
+        this.performingBlobStore = performingBlobStore;
+    }
+
+    @Override
+    public Mono<BlobId> save(BucketName bucketName, byte[] data, StoragePolicy storagePolicy) {
+        return selectBlobStore(storagePolicy, () -> data.length > SIZE_THRESHOLD)
+            .save(bucketName, data, storagePolicy);
```

Comment:
you didn't fallback to the another if fail?

Reply:
No.

Storing a 250MB stream in Cassandra will harm overall server performance that's not a risk I wonna take.